src/Behat/Testwork/ServiceContainer/ExtensionManager.php

```diff
@@ -34,30 +32,50 @@
     private $debugInformation = array(
         'extensions_list' => array()
     );
+    /**
+     * @var ExtensionInstantiator[]
+     */
+    private $instantiators = array();
 
     /**
      * Initializes manager.
      *
-     * @param Extension[] $extensions     List of default extensions
-     * @param null|string $extensionsPath Base path where to search custom extension files
+     * @param Extension[]             $extensions      List of default extensions
+     * @param string|null             $extensionsPath  Base path where to search custom extension files (deprecated)
+     * @param ExtensionInstantiator[] $instantiators   Extension instantiators to use
      */
-    public function __construct(array $extensions, $extensionsPath = null)
+    public function __construct(array $extensions, $extensionsPath = null, array $instantiators = array())
     {
         foreach ($extensions as $extension) {
             $this->extensions[$extension->getConfigKey()] = $extension;
         }
 
-        $this->extensionsPath = $extensionsPath;
+        // BC Layer for `extensionsPath`
+        // to be removed in 4.0
+        if (empty($instantiators)) {
+            $instantiators = array(
+                new ExtensionClassInstantiator(),
+                new ExtensionFileInstantiator($extensionsPath),
+            );
+        }
+
+        $this->instantiators = $instantiators;
     }
 
     /**
      * Sets path to directory in which manager will try to find extension files.
      *
      * @param null|string $path
+     * @deprecated since 3.4, to be removed in 4.0
+     *  (Set the extensions path through the proper instantiator instead)
      */
     public function setExtensionsPath($path)
     {
-        $this->extensionsPath = $path;
+        foreach ($this->instantiators as $instantiator) {
+            if ($instantiator instanceof ExtensionFileInstantiator) {
```

Comment:
instead of testing if the instantiator is the `ExtensionFileInstantiator`, maybe testing if the instantiator has a method `setExtensionsPath` would be better ?